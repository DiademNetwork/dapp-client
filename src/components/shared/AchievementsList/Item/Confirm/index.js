import React, { Component} from 'react'
import * as R from 'ramda'
import network from 'configurables/network'
import { PropTypes as T } from 'prop-types'
import Typography from '@material-ui/core/Typography'
import DialogContentText from '@material-ui/core/DialogContentText'
import withContainer from './container'
import Modal from 'components/shared/Modal'
import Link from 'components/shared/Link'
import UserName from 'components/shared/UserName'

class ConfirmAchievement extends Component {
  constructor (props) {
    super(props)
    this.state = {
      hasConfirmedAlready: this.hasUserAlreadyConfirmed(props.achievement),
      creatorAddress: R.prop('id')(this.getCreationActivity(props.achievement.activities))
    }
  }

  hasUserAlreadyConfirmed = () => false

  handleConfirm = () => {
    const { accessToken, confirmAchievement, achievement, userID, userAddress } = this.props
    const { creatorAddress } = this.state
    confirmAchievement({
      userAddress: userAddress,
      creatorAddress,
      link: achievement.group,
      token: accessToken,
      user: userID
    })
  }

  getCreationActivity = R.compose(
    R.head,
    R.filter(R.propEq('verb', 'create'))
  )

  render () {
    const {
      achievement,
      canPerformActions,
      idx,
      link,
      title
    } = this.props
    const { hasConfirmedAlready } = this.state
    const { actor } = this.getCreationActivity(achievement.activities)
    return (
      <Modal
        confirmButtonText="Confirm"
        data-qa-id={`achievement-${idx}-confirm-modal`}
        disabled={hasConfirmedAlready || !canPerformActions}
        name={`achievement-${idx}-confirm-button`}
        onConfirm={this.handleConfirm}
        openButtonText={hasConfirmedAlready ? 'You confirmed already' : 'Confirm'}
        title="Confirm"
        render={() => (
          <DialogContentText component="div">
            <Typography paragraph variant="body1">
              Has <UserName actor={actor} /> really done achievement?<br /><br />
              {title}
            </Typography>
            <Link
              text={`View achievement post on ${network.name} again`}
              href={link}
              typographyProps={{
                paragraph: true
              }}
            />
            <Typography variant="caption" color="textSecondary">
              Confirmations make other Diadem Network users know they support real achievement(s)<br />
              Please confirm only achievement(s) you are sure of
            </Typography>
          </DialogContentText>
        )}
      />
    )
  }
}

ConfirmAchievement.propTypes = {
  canPerformActions: T.bool,
  achievement: T.object,
  accessToken: T.string,
  confirmAchievement: T.func,
  idx: T.number,
  link: T.string,
  title: T.string,
  userID: T.string,
  userAddress: T.string
}

export default R.compose(
  withContainer
)(ConfirmAchievement)
