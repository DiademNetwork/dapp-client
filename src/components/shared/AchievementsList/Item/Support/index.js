import React from 'react'

const AchievementListItemSupport = () => <span>support</span>

export default AchievementListItemSupport
