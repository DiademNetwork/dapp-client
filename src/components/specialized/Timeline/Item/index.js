import React from 'react'
import { PropTypes as T } from 'prop-types'
import UserName from 'components/shared/UserName'
import UserAvatar from 'components/shared/UserAvatar'
import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'
import moment from 'moment'

const styles = (theme) => ({
  link: {
    color: theme.palette.primary.light,
    textDecoration: 'none'
  }
})

const TimelineItem = ({ classes, item: { actor, verb, object, time, title } }) => (
  <ListItem>
    <UserAvatar actor={actor} />
    <ListItemText primary={[
      <Typography color="textSecondary" key='time'>{moment.utc(time).from(moment.utc(new Date()))}</Typography>,
      (function getMainText () {
        const achievementLink = <a className={classes.link} href={object} target="_blank">{title || 'achivement'}</a>
        switch (verb) {
          case 'create':
            return (<Typography key='text'><UserName actor={actor} /> created {achievementLink}</Typography>)
          case 'support':
            return (<Typography key='text'><UserName actor={actor} /> supported {achievementLink}</Typography>)
          case 'confirm':
            return (<Typography key='text'><UserName actor={actor} /> confirmed {achievementLink}</Typography>)
          default:
            return null
        }
      })()
    ]} />
  </ListItem>
)

TimelineItem.propTypes = {
  classes: T.object,
  item: T.object
}

export default withStyles(styles)(TimelineItem)
