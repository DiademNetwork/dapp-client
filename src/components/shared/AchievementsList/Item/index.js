import React, { Fragment } from 'react'
import * as R from 'ramda'
import { PropTypes as T } from 'prop-types'
import Card from '@material-ui/core/Card'
import CardHeader from '@material-ui/core/CardHeader'
import CardContent from '@material-ui/core/CardContent'
import CardActions from '@material-ui/core/CardActions'
import Divider from '@material-ui/core/Divider'
import Link from 'components/shared/Link'
import UserName from 'components/shared/UserName'
import UserAvatar from 'components/shared/UserAvatar'
import { withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import Confirm from './Confirm'
import Support from './Support'
import withContainer from './container'

const styles = (theme) => ({
  item: {
    marginBottom: theme.spacing.unit * 2
  },
  actions: {
    justifyContent: 'flex-end'
  },
  link: {
    marginBottom: theme.spacing.unit
  }
})

const getCreationActivity = R.compose(
  R.head,
  R.filter(R.propEq('verb', 'create'))
)

const getConfirmActivities = R.compose(
  R.filter(R.propEq('verb', 'confirm'))
)

const AchievementsListItem = ({
  classes,
  item,
  idx,
  userAddress
}) => {
  const { activities } = item
  const creationActivity = getCreationActivity(activities)
  const { actor: creator, title, object } = creationActivity
  return (
    <Card
      className={classes.item}
      data-qa-id={`achievement-item-${idx}`}
      key="achievement-card"
    >
      <CardHeader
        avatar={<UserAvatar actor={creator} />}
        title={
          <Typography key="achievement-title" variant="headline">
            <Link
              className={classes.link}
              href={object}
              text={title}
              variant="headline"
            />
          </Typography>
        }
        subheader={<Typography key="achievement-creator" variant="subheading" color="textSecondary">by <UserName actor={creator} /></Typography>}
      />
      <Divider />
      <CardContent>
        This achievement has been confirmed {getConfirmActivities(activities).length}
      </CardContent>
      <CardActions
        className={classes.actions}
        disableActionSpacing
      >
        {UserName.getUserAddress(creator) !== userAddress ? (
          <Fragment>
            <Confirm key='confirm' achievement={item} />
            <Support key='support' />
          </Fragment>
        ) : (
          <Typography>This is your achievement</Typography>
        )}
      </CardActions>
    </Card>
  )
}

AchievementsListItem.propTypes = {
  classes: T.object,
  idx: T.number,
  item: T.object,
  userAddress: T.string
}

export default R.compose(
  withContainer,
  withStyles(styles)
)(AchievementsListItem)
