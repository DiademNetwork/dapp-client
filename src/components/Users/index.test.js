import React from 'react'
import { shallow } from 'enzyme'
import { Component as Users } from './index'
import Item from './Item'

const users = [{ userName: 'A' }, { userName: 'B' }]

describe('Users', () => {
  describe('With users', () => {
    it('renders without crashing', () => {
      const component = shallow(<Users users={users} />)
      expect(component).toMatchSnapshot()
    })

    it('As many Users items as users in list rendered', () => {
      const component = shallow(<Users users={users} />)
      expect(component.find(Item).length).toEqual(users.length)
    })
  })
  describe('Without users', () => {
    it('renders without crashing', () => {
      const component = shallow(<Users />)
      expect(component).toMatchSnapshot()
    })

    it('Loading when no users and users being requested', () => {
      const component = shallow(<Users fetchStatus='requested' users={[]} />)
      expect(component.find(`[data-qa-id='no-user-text']`).text()).toEqual('Loading...')
    })

    it('No user tecxt when no users and users not being requested', () => {
      const component = shallow(<Users users={[]} />)
      expect(component.find(`[data-qa-id='no-user-text']`).text()).toEqual('No user registered')
    })
  })
})
