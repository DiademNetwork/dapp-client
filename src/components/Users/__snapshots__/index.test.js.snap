// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users With users renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Users
    users={
      Array [
        Object {
          "userName": "A",
        },
        Object {
          "userName": "B",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <WithStyles(CardContent)>
        <WithStyles(Typography)
          color="textSecondary"
          paragraph={true}
        >
          All registered Diadem Network users
        </WithStyles(Typography)>
        <WithStyles(List)>
          <UserItem
            user={
              Object {
                "userName": "A",
              }
            }
          />
          <UserItem
            user={
              Object {
                "userName": "B",
              }
            }
          />
        </WithStyles(List)>
      </WithStyles(CardContent)>,
      "className": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Typography)
            color="textSecondary"
            paragraph={true}
          >
            All registered Diadem Network users
          </WithStyles(Typography)>,
          <WithStyles(List)>
            <UserItem
              user={
                Object {
                  "userName": "A",
                }
              }
            />
            <UserItem
              user={
                Object {
                  "userName": "B",
                }
              }
            />
          </WithStyles(List)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "All registered Diadem Network users",
            "color": "textSecondary",
            "paragraph": true,
          },
          "ref": null,
          "rendered": "All registered Diadem Network users",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <UserItem
                user={
                  Object {
                    "userName": "A",
                  }
                }
              />,
              <UserItem
                user={
                  Object {
                    "userName": "B",
                  }
                }
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "function",
              "props": Object {
                "user": Object {
                  "userName": "A",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "function",
              "props": Object {
                "user": Object {
                  "userName": "B",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <WithStyles(CardContent)>
          <WithStyles(Typography)
            color="textSecondary"
            paragraph={true}
          >
            All registered Diadem Network users
          </WithStyles(Typography)>
          <WithStyles(List)>
            <UserItem
              user={
                Object {
                  "userName": "A",
                }
              }
            />
            <UserItem
              user={
                Object {
                  "userName": "B",
                }
              }
            />
          </WithStyles(List)>
        </WithStyles(CardContent)>,
        "className": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(Typography)
              color="textSecondary"
              paragraph={true}
            >
              All registered Diadem Network users
            </WithStyles(Typography)>,
            <WithStyles(List)>
              <UserItem
                user={
                  Object {
                    "userName": "A",
                  }
                }
              />
              <UserItem
                user={
                  Object {
                    "userName": "B",
                  }
                }
              />
            </WithStyles(List)>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "All registered Diadem Network users",
              "color": "textSecondary",
              "paragraph": true,
            },
            "ref": null,
            "rendered": "All registered Diadem Network users",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <UserItem
                  user={
                    Object {
                      "userName": "A",
                    }
                  }
                />,
                <UserItem
                  user={
                    Object {
                      "userName": "B",
                    }
                  }
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "function",
                "props": Object {
                  "user": Object {
                    "userName": "A",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "function",
                "props": Object {
                  "user": Object {
                    "userName": "B",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`Users Without users renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Users
    users={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <WithStyles(CardContent)>
        <WithStyles(Typography)
          color="textSecondary"
          paragraph={true}
        >
          All registered Diadem Network users
        </WithStyles(Typography)>
        <WithStyles(Typography)
          color="textPrimary"
        >
          <span
            data-qa-id="no-user-text"
          >
            No user registered
          </span>
        </WithStyles(Typography)>
      </WithStyles(CardContent)>,
      "className": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Typography)
            color="textSecondary"
            paragraph={true}
          >
            All registered Diadem Network users
          </WithStyles(Typography)>,
          <WithStyles(Typography)
            color="textPrimary"
          >
            <span
              data-qa-id="no-user-text"
            >
              No user registered
            </span>
          </WithStyles(Typography)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "All registered Diadem Network users",
            "color": "textSecondary",
            "paragraph": true,
          },
          "ref": null,
          "rendered": "All registered Diadem Network users",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <span
              data-qa-id="no-user-text"
            >
              No user registered
            </span>,
            "color": "textPrimary",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "No user registered",
              "data-qa-id": "no-user-text",
            },
            "ref": null,
            "rendered": "No user registered",
            "type": "span",
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <WithStyles(CardContent)>
          <WithStyles(Typography)
            color="textSecondary"
            paragraph={true}
          >
            All registered Diadem Network users
          </WithStyles(Typography)>
          <WithStyles(Typography)
            color="textPrimary"
          >
            <span
              data-qa-id="no-user-text"
            >
              No user registered
            </span>
          </WithStyles(Typography)>
        </WithStyles(CardContent)>,
        "className": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(Typography)
              color="textSecondary"
              paragraph={true}
            >
              All registered Diadem Network users
            </WithStyles(Typography)>,
            <WithStyles(Typography)
              color="textPrimary"
            >
              <span
                data-qa-id="no-user-text"
              >
                No user registered
              </span>
            </WithStyles(Typography)>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "All registered Diadem Network users",
              "color": "textSecondary",
              "paragraph": true,
            },
            "ref": null,
            "rendered": "All registered Diadem Network users",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <span
                data-qa-id="no-user-text"
              >
                No user registered
              </span>,
              "color": "textPrimary",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "No user registered",
                "data-qa-id": "no-user-text",
              },
              "ref": null,
              "rendered": "No user registered",
              "type": "span",
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
